Update documentation of gpu section (#789)
intel/scikit-learn-intelex: doc/sources/oneapi_gpu.rst

.. *******************************************************************************/

.. _oneapi_gpu:

##############################################################
oneAPI and GPU support in Intel(R) Extension for Scikit-learn*
##############################################################

Intel(R) Extension for Scikit-learn* provides support of oneAPI concepts, which
means that algorithms can be executed on different devices - CPUs and GPUs.
This is done by integration with
`dpctl <https://intelpython.github.io/dpctl/latest/index.html>`_ package, that
implements core of oneAPI concepts like queues and devices.

Intel(R) Extension for Scikit-learn* offers two ways how to run an algorithm on
specific device with help of dpctl:

1. Pass input data as
   `dpctl.tensor.usm_ndarray
   <https://intelpython.github.io/dpctl/latest/docfiles/dpctl.tensor_api.html#dpctl.tensor.usm_ndarray>`_
   to the algorithm. The computation will run on the device where input data is
   located, and result will be returned as :code:`usm_ndarray` on the same
   device.

   .. note::
     All the data passed to the algorithm must reside on the same device.

   .. warning::

>      Only base methods like :code:`fit`, :code:`predict`, and :code:`transform`
>      can consume :code:`usm_ndarray`. Please pay attension that only the
>      algorithms in Intel(R) Extension for Scikit-learn* provide support of
>      :code:`usm_ndarray`. Standard algorithms in stock version of Scikit-learn*
>      do not include this feature.

Reviewer: ```suggestion
     The :code:`usm_ndarray` can only be consumed by the base methods
     like :code:`fit`, :code:`predict`, and :code:`transform`.
     Note that only the algorithms in Intel(R) Extension for Scikit-learn* support
     :code:`usm_ndarray`. The algorithms from the stock version of scikit-learn
     do not support this feature.
```

2. Use global configurations of Intel(R) Extension for Scikit-learn*:

>      1. Option :code:`target_offload` can be used to set the device primarily
>         used to perform computations. Accepted values are :code:`str` and
>         :code:`dpctl.SyclQueue`. If string, expected to be "auto" (the execution
>         context is deduced from input data location), or SYCL* filter selector
>         string. Default value is "auto"
>      2. Option :code:`allow_fallback_to_host`
>         is a boolean flag, that, if set, allows to fallback computation to host
>         device in case when particular estimator does not support the selected
>         one. Default value is :code:`False`.

Reviewer: `Accepted values` -- you really mean `accepted data types`, no?

I also don't think my changes below are correct, because it's not clear what happens if the input is dpctl.SyclQueue.

```suggestion
     1. The :code:`target_offload` option can be used to set the device primarily
        used to perform computations. Accepted values are :code:`str` and
        :code:`dpctl.SyclQueue`. The default value is "auto", meaning that the execution
        context is deduced from the location of input data. Otherwise, a string with a
        SYCL* filter selector is expected.
     2. The :code:`allow_fallback_to_host` option
        is a Boolean flag. If set to :code:`True`, the computation is allowed to fallback to the host
        device when a particular estimator does not support the selected
        device. The default value is :code:`False`.
```

Reviewer 2: `I also don't think my changes below are correct, because it's not clear what happens if the input is dpctl.SyclQueue.`
It means that if the data type is a string, then it can take the values "auto" or SYCL * filter selector. For the dpctl type.SyclQueue there are no restrictions

These options can be set using :code:`sklearnex.set_config()` function or
:code:`sklearnex.config_context`. Current values of these options can be
obtained by the call of :code:`sklearnex.get_config()`.

.. note::
     Functions :code:`set_config`, :code:`get_config` and :code:`config_context`

>      are always patched after :code:`sklearnex.patch_sklearn()` call.

Reviewer: ```suggestion
     are always patched after the :code:`sklearnex.patch_sklearn()` call.
```

> Example how to patch your code with Intel CPU/GPU optimizations:

Reviewer: ```suggestion
.. rubric:: Example

An example on how to patch your code with Intel CPU/GPU optimizations:
```

or

```suggestion
Example
#######

An example on how to patch your code with Intel CPU/GPU optimizations:
```

.. code-block:: python

   from sklearnex import patch_sklearn, config_context
   patch_sklearn()

   from sklearn.cluster import DBSCAN

   X = np.array([[1., 2.], [2., 2.], [2., 3.],
               [8., 7.], [8., 8.], [25., 80.]], dtype=np.float32)
   with config_context(target_offload="gpu:0"):
      clustering = DBSCAN(eps=3, min_samples=2).fit(X)

.. warning::
     For compatibility reasons, algorithms in Intel(R) Extension for
     Scikit-learn* can be offloaded to the device with use of
     :code:`daal4py.oneapi.sycl_context`. It is recommended to use the ways

>      described above for device offloading instead of using :code:`sycl_context`.

Reviewer: I think this should be placed before the example. I also don't think this is a warning, you may place this in the main text body. You may use a section header (e.g. "Compatibility considerations" or something similar) to make it stand out.

I also have questions about this paragraph.
- `compatibility reasons` -- compatibility with what? daal4py? something else?
- are you saying that the user may use sycl_context if they have to, but we do not recommend this?

Reviewer 2: * `compatibility reasons -- compatibility with what? daal4py? something else?` 
I will clarify this question and formulate the reason more precisely
* `are you saying that the user may use sycl_context if they have to, but we do not recommend this?`
Yes, we still have sycl_context support, but it is preferable to use the methods suggested in the device offloading paragraph

Reviewer 2: `Compatibility reasons`  - compatibility with users scripts. You can use `sycl_context`, but it's not recommended.

For execution on GPU, DPC++ compiler runtime and driver are required. Refer to `DPC++ system
requirements <https://software.intel.com/content/www/us/en/develop/articles/intel-oneapi-dpcpp-system-requirements.html>`_ for details.

DPC++ compiler runtime can be installed either from PyPI or Anaconda:

- Install from PyPI::

     pip install dpcpp-cpp-rt

- Install from Anaconda::

>      conda install dpcpp_cpp_rt -c intel

Reviewer: lines 89-100 should be placed at the beginning of this topic and clearly marked as "requirements" or "prerequisites"

     conda install dpcpp_cpp_rt -c intel

Device offloading
-----------------

Intel(R) Extension for Scikit-learn* offers two options for running an algorithm on a
specific device with the help of dpctl:

1. Pass input data as `dpctl.tensor.usm_ndarray <https://intelpython.github.io/dpctl/latest/docfiles/dpctl.tensor_api.html#dpctl.tensor.usm_ndarray>`_ to the algorithm.

   The computation will run on the device where the input data is
   located, and the result will be returned as :code:`usm_ndarray` to the same
   device.

   .. note::
     All the input data for an algorithm must reside on the same device.

   .. warning::
     The :code:`usm_ndarray` can only be consumed by the base methods
     like :code:`fit`, :code:`predict`, and :code:`transform`.
     Note that only the algorithms in Intel(R) Extension for Scikit-learn* support
     :code:`usm_ndarray`. The algorithms from the stock version of scikit-learn
     do not support this feature.
2. Use global configurations of Intel(R) Extension for Scikit-learn\*:
     1. The :code:`target_offload` option can be used to set the device primarily
        used to perform computations. Accepted data types are :code:`str` and

>         :code:`dpctl.SyclQueue`. If string, expected to be "auto" (the execution
>         context is deduced from input data location), or SYCL* filter selector
>         string. Default value is "auto"

Reviewer: ```suggestion
        :code:`dpctl.SyclQueue`. If you pass a string to :code:`target_offload`,
        it should either be ``"auto"``, which means that the execution
        context is deduced from the location of input data, or a string
        with SYCL* filter selector. The default value is ``"auto"``.
```

Reviewer: @vlad-nazarov please take a look at this wording

Reviewer 2: @outoftardis, this wording looks good to me

     2. The :code:`allow_fallback_to_host` option
        is a Boolean flag. If set to :code:`True`, the computation is allowed 
        to fallback to the host device when a particular estimator does not support
        the selected device. The default value is :code:`False`.

These options can be set using :code:`sklearnex.set_config()` function or
:code:`sklearnex.config_context`. To obtain the current values of these options,
call :code:`sklearnex.get_config()`.

.. note::
     Functions :code:`set_config`, :code:`get_config` and :code:`config_context`
     are always patched after the :code:`sklearnex.patch_sklearn()` call.

> **Compatibility considerations**::

Reviewer: ```suggestion
.. rubric:: Compatibility considerations
```

For compatibility reasons, algorithms in Intel(R) Extension for

> Scikit-learn* can be offloaded to the device with use of

Reviewer: ```suggestion
Scikit-learn* may be offloaded to the device using
```

> :code:`daal4py.oneapi.sycl_context`. It is recommended to use the ways
> described above for device offloading instead of using :code:`sycl_context`.

Reviewer: ```suggestion
:code:`daal4py.oneapi.sycl_context`. However, it is recommended to use one of the options
described above for device offloading instead of using :code:`sycl_context`.
```
